src/Interpreters/InterpreterInsertQuery.cpp

```diff
@@ -256,27 +261,35 @@ BlockIO InterpreterInsertQuery::execute()
     }
 
     /// What type of query: INSERT or INSERT SELECT or INSERT WATCH?
-    if (query.select || query.watch)
+    if (is_distributed_insert_select)
+    {
+        /// Pipeline was already built.
+    }
+    else if (query.select || query.watch)
     {
-        for (auto & in_stream : in_streams)
+        const auto & header = out_streams.at(0)->getHeader();
+
+        res.pipeline.addSimpleTransform([&](const Block & in_header) -> ProcessorPtr
         {
-            in_stream = std::make_shared<ConvertingBlockInputStream>(
-                in_stream, out_streams.at(0)->getHeader(), ConvertingBlockInputStream::MatchColumnsMode::Position);
-        }
+            return std::make_shared<ConvertingTransform>(in_header, header,
+                    ConvertingTransform::MatchColumnsMode::Position);
+        });
 
-        Block in_header = in_streams.at(0)->getHeader();
-        if (in_streams.size() > 1)
+        res.pipeline.setSinks([&](const Block &, QueryPipeline::StreamType type) -> ProcessorPtr
         {
-            for (size_t i = 1; i < in_streams.size(); ++i)
-                assertBlocksHaveEqualStructure(in_streams[i]->getHeader(), in_header, query.select ? "INSERT SELECT" : "INSERT WATCH");
-        }
+            if (type != QueryPipeline::StreamType::Main)
+                return nullptr;
 
-        res.in = std::make_shared<NullAndDoCopyBlockInputStream>(in_streams, out_streams);
+            auto stream = std::move(out_streams.back());
```

Comment:
It seems I understand how it worked previously, but I don't understand these changes :(

Reply:
The idea is that we wrap each output stream to processor (sink), and add them to pipeline. It looks like:
```
-input-->        -> sink
-input--> Resize -> sink
-input-->        -> sink
```